### ISSUE TITLE
using set_tabular_sfh 

### ISSUE THREAD

**Issue by dptriani** [dfm/python-fsps#91]
Hi,
I am going to use set_tabular_sfh option but I test it first. So I am running this script:

sp = fsps.StellarPopulation()
sdss_bands = fsps.find_filter('sdss')
mags = sp.get_mags(bands=sdss_bands)
age = (10 ** sp.log_age) * (10 ** (-9))
sfr = sp.sfr
wavelength, spectrum = sp.get_spectrum(tage=13.7)

sp1 = fsps.StellarPopulation(sfh=3)
sp1.set_tabular_sfh(age=age, sfr=sfr)
wavelength1, spectrum1 = sp1.get_spectrum(tage=13.7)

and hoping that spectrum1 and spectrum would give similar result but no, the results differ on order 20. Why is this?

**Comment by zpace** [comment]
At some point, I was advised to avoid instantiating `fsps.StellarPopulation` more than once: the maintainers would be better equipped to answer precisely why, but I suspect it has to do with how these python bindings interface with the underlying fortran.

Try running your two code blocks as separate programs, write out the result for each, and compare them that way.

**Comment by dptriani** [comment]
I have separated them but they still give difference result

**Comment by bd-j** [comment]
It's true that to be safe you want to have a single StellarPopulation object.  But the issue here I think is more that the default SFH (used in your first `sp` object) is a single age simple stellar population (`sfh=0`).  The tabular `sp.sfr` value is not actually well-defined for such an SFH (since it is a delta function at age=0) and so the `sfr` variable that you compute is filled with zeros and the resulting `sfh=3` spectrum is garbage.  I guess we could check for non-zero sfr inside `set_tabular_sfh()` to warn users about this.

What I'd do instead  is
```python
import fsps
import matplotlib.pyplot as pl

sps = fsps.StellarPopulation()

# get the spectrum via the intrinsic SFH convolutions
sps.params["sfh"] = 4  #delayed_exponential
sdss_bands = fsps.find_filter("sdss")
mags = sps.get_mags(bands=sdss_bands)
age = (10 ** sps.log_age) / 1e9
sfr = sps.sfr
wavelength, spectrum = sps.get_spectrum(tage=13.7)

# get the spectrum from a tabular version
sps.params["sfh"] = 3
sps.set_tabular_sfh(age, sfr)
wavelength, spectrum1 = sps.get_spectrum(tage=13.7)

# Plot the SFR
fig, axes = pl.subplots(2, 1)
axes[0].plot(age, sfr, '-o')
axes[0].set_xlabel('Age (Gyr)')
axes[0].set_ylabel('SFR (M$_\odot$/yr')
axes[1].plot(wavelength, spectrum, label="via sfh=4")
axes[1].plot(wavelength, spectrum1, label="via sfh=3")
axes[1].set_xlim(1e3, 1e4)
axes[1].set_xlabel('$\lambda (\AA)$')
axes[1].set_ylabel('$f_\nu (L_\odot/Hz)$')
axes[1].legend(loc=0)
pl.show()
```

In fact you will probably still notice (smaller) differences.  I'm pretty sure this is because for `sfh=1, 4, or 5` all the calculations are done exactly, taking into the account the true exponential dependence of the SFR between ssp ages, whereas for `sfh=3` the SFH is assumed to be piecewise linear between the provided tabular age points, which becomes a poor approximation to the true SFH as the age resolution decreases.  But it could be good to do some more testing of that.

**Comment by dptriani** [comment]
Thanks a lot for the answer!

**Comment by bd-j** [comment]
Is this with `sfh=3`?  I'm not sure how you'd do a burst that way.  Also, by tstart do you mean the `sf_start` parameter?  Or the `tburst` parameter?  Both of those are only defined for `sfh=1,4,5`

**Comment by bd-j** [comment]
Yes, the times (`sf_start`, `tburst`, `tage`, `log_age`) are age of universe, not lookback time.

**Comment by bd-j** [comment]
Just wanted to add a note here that the difference between the spectra obtained via SFH=3 and SFH=4 for delay-tau that I mentioned earlier in the thread was actually caused largely by an FSPS bug in tabular SFH calculations, now fixed by cconroy20/fsps#32

Remaining differences are much less than a percent for lambda > 1300 AA, and less than 5 percent blueward of that.
